pkg/validate/container_linux.go

```diff
@@ -307,3 +307,201 @@ func createOOMKilledContainer(
+			if rro && !runtimeSupportsRRO(rc, runtimeHandler) {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+
+			By("create host path")
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+
+			By("create container with volume")
+			containerID := createRROMountContainer(rc, ic, podID, podConfig, hostPath, "/mnt", rro)
+
+			By("test start container with volume")
+			testStartContainer(rc, containerID)
+
+			By("check whether `touch /mnt/tmpfs/file` succeeds")
+			command := []string{"touch", "/mnt/tmpfs/file"}
+			if rro {
+				command = []string{"sh", "-c", `touch /mnt/tmpfs/foo 2>&1 | grep -q "Read-only file system"`}
+			}
+			execSyncContainer(rc, containerID, command)
+		}
+
+		It("should support non-recursive readonly mounts", func() {
+			testRRO(rc, ic, false)
+		})
+		It("should support recursive readonly mounts", func() {
+			testRRO(rc, ic, true)
+		})
+		testRROInvalidPropagation := func(prop runtimeapi.MountPropagation) {
+			if !runtimeSupportsRRO(rc, runtimeHandler) {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+			mounts := []*runtimeapi.Mount{
+				{
+					HostPath:          hostPath,
+					ContainerPath:     "/mnt",
+					Readonly:          true,
+					RecursiveReadOnly: true,
+					SelinuxRelabel:    true,
+					Propagation:       prop,
+				},
+			}
+			const expectErr = true
+			createMountContainer(rc, ic, podID, podConfig, mounts, expectErr)
+		}
+		It("should reject a recursive readonly mount with PROPAGATION_HOST_TO_CONTAINER", func() {
+			testRROInvalidPropagation(runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER)
+		})
+		It("should reject a recursive readonly mount with PROPAGATION_BIDIRECTIONAL", func() {
+			testRROInvalidPropagation(runtimeapi.MountPropagation_PROPAGATION_BIDIRECTIONAL)
+		})
+		It("should reject a recursive readonly mount with ReadOnly: false", func() {
+			if !runtimeSupportsRRO(rc, runtimeHandler) {
+				Skip("runtime does not implement recursive readonly mounts")
+				return
+			}
+			hostPath, clearHostPath := createHostPathForRROMount(podID)
+			defer clearHostPath() // clean up the TempDir
+			mounts := []*runtimeapi.Mount{
+				{
+					HostPath:          hostPath,
+					ContainerPath:     "/mnt",
+					Readonly:          false,
+					RecursiveReadOnly: true,
+					SelinuxRelabel:    true,
+				},
+			}
+			const expectErr = true
+			createMountContainer(rc, ic, podID, podConfig, mounts, expectErr)
+		})
+	})
+})
+
+func runtimeSupportsRRO(rc internalapi.RuntimeService, runtimeHandlerName string) bool {
+	ctx := context.Background()
+	status, err := rc.Status(ctx, false)
+	framework.ExpectNoError(err, "failed to check runtime status")
+	for _, h := range status.RuntimeHandlers {
+		if h.Name == runtimeHandlerName {
```

Comment:
If no runtime handler set (`""`), do we then want to pick the first one supporting the RRO mounts? We did the same for the user namespaces: https://github.com/kubernetes-sigs/cri-tools/blob/3a1f73d79097afc512a841aae94dae20704e9eb4/pkg/validate/security_context_linux.go#L861-L876

Reply:
I'd say userns test should just use the default handler for better determinism

Reply:
Makes sense, I'll align the behvario for the user namespaces tests in https://github.com/kubernetes-sigs/cri-tools/pull/1369.